File: src/main/java/fr/gouv/tac/analytics/server/service/AnalyticsService.java
package fr.gouv.tac.analytics.server.service;

import fr.gouv.tac.analytics.server.config.AnalyticsProperties;
import fr.gouv.tac.analytics.server.service.kafka.model.AnalyticsCreation;
import fr.gouv.tac.analytics.server.service.kafka.model.AnalyticsDeletion;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class AnalyticsService {

    private final AnalyticsProperties analyticsProperties;

    private final KafkaTemplate<String, Object> kafkaTemplate;

    public void createAnalytics(final AnalyticsCreation analytics) {
        kafkaTemplate.send(analyticsProperties.getCreationTopic(), analytics)
                .addCallback(
                        sendResult -> log.debug("Message successfully sent {}", sendResult),
                        throwable -> log.warn("Analytics creation - error sending message to kafka", throwable)
                );
    }

    public void deleteAnalytics(final String installationUuid) {
        final var analyticsDeletion = AnalyticsDeletion.builder()
                .installationUuid(installationUuid)
                .deletionTimestamp(Instant.now())
                .build();

        kafkaTemplate.send(analyticsProperties.getDeletionTopic(), analyticsDeletion)
                .addCallback(
                        sendResult -> log.debug("Message successfully sent {}", sendResult),
                        throwable -> log.warn("Analytics deletion - error sending message to kafka", throwable)
                );
    }
}

File: src/main/kotlin/fr/gouv/tac/analytics/service/AnalyticsService.kt
package fr.gouv.tac.analytics.service

import fr.gouv.tac.analytics.config.AnalyticsProperties
import fr.gouv.tac.analytics.model.AnalyticsCreation
import fr.gouv.tac.analytics.model.AnalyticsDeletion
import org.slf4j.LoggerFactory
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.stereotype.Service
import java.time.Instant.now
import java.time.OffsetDateTime

@Service
class AnalyticsService(
    private val analyticsProperties: AnalyticsProperties,
    private val kafkaTemplate: KafkaTemplate<String, Any>
) {
    private val log = LoggerFactory.getLogger(AnalyticsService::class.java)

    fun createAnalytics(analytics: AnalyticsCreation) {
        val analyticsEvent = analytics.copy(creationDate = OffsetDateTime.now())
        kafkaTemplate.send(analyticsProperties.creationTopic, analyticsEvent)
            .addCallback(
                { log.debug("Message successfully sent {}", it) },
                { log.warn("Analytics creation - error sending message to kafka", it) }
            )
    }

    fun deleteAnalytics(installationUuid: String) {
        val analyticsDeletion = AnalyticsDeletion(
            installationUuid = installationUuid,
            deletionTimestamp = now()
        )
        kafkaTemplate.send(analyticsProperties.deletionTopic, analyticsDeletion)
            .addCallback(
                { log.debug("Message successfully sent {}", it) },
                { log.warn("Analytics deletion - error sending message to kafka", it) }
            )
    }
}